torch/csrc/distributed/rpc/rref.h

```diff
@@ -0,0 +1,137 @@
+// will *eventually* get to the peer,  and that the peer will create a RRef out
+// of it. Therefore, no constructor of RRefForkData is exposed, and
+// applications should never directly use RRefForkData. All construction are
+// done within ``RRef`` and ``RRefContext``.
+struct RRefForkData {
+  at::IValue toIValue() const;
+ private:
+  friend class RRef;
+  friend class RRefContext;
+  friend class UserRRef;
+
+  RRefForkData(worker_id_t ownerId,
+               const RRefId& rrefId_,
+               const ForkId& forkId_);
+
+  static RRefForkData fromIValue(at::IValue&&);
+
+  const worker_id_t ownerId_;
+  const RRefId rrefId_;
+  const ForkId forkId_;
+};
+
+// TODO: make RRef an IValue, and edit createStackForSchema accordingly
+class RRef {
+ public:
+  // RRef is made NOT copyable NOT movable to prevent messing up reference
+  // counting
+  RRef(const RRef& other) = delete;
+  RRef(RRef&& other) = delete;
+
+  virtual ~RRef() = default;
+
+  worker_id_t owner() const;
+  const RRefId& id() const;
+  IValue fork() const;
+
+  virtual bool isOwner() const = 0;
+  virtual IValue toHere() = 0;
+
+ protected:
+  RRef(worker_id_t ownerId, const RRefId& rrefId);
+
+  const worker_id_t ownerId_;
+  const RRefId rrefId_;
+};
+
+class UserRRef final: public RRef {
+ public:
+  const ForkId& forkId() const;
+  bool isOwner() const override;
+  IValue toHere() override;
+
+  ~UserRRef() override;
+ private:
+  friend class RRefContext;
+
+  UserRRef(worker_id_t ownerId, const RRefId& rrefId, const ForkId& forkId);
+
+  const ForkId forkId_;
+};
+
+// Keep the template only on the derived class because ``RRefContext`` needs to
+// erase the type on ``RRef`` and keep them in one map.
+template <typename T>
+class OwnerRRef final: public RRef {
+ public:
+  bool isOwner() const override {
+    return true;
+  }
+
+  T getValue() const {
+    // TODO: use callback to make this non-blocking
+    std::unique_lock<std::mutex> lock(mutex_);
+    valueCV_.wait(lock, [this]{return value_.has_value();});
+    return value_.value();
+  }
+
+  void setValue(T&& value) {
+    {
+      std::lock_guard<std::mutex> lock(mutex_);
+      value_ = std::move(value);
+    }
+    valueCV_.notify_all();
+  }
+
+  IValue toHere() override {
+    AT_ERROR("OwnerRRef does not support toHere(), use getValue() instead.");
+  }
+
+  // TODO: add setValue(py::object) and getPyObj() for Python UDF
+
+ private:
+  friend class RRefContext;
+
+  OwnerRRef(worker_id_t ownerId, const RRefId& rrefId)
+      : OwnerRRef(ownerId, rrefId, {}) {}
+
+  OwnerRRef(OwnerRRef<T>&& other) noexcept
+      : OwnerRRef(other.owner(), other.id(), std::move(other.value_)) {}
+
+  OwnerRRef(worker_id_t ownerId, const RRefId& rrefId, c10::optional<T> value)
+      : RRef(ownerId, rrefId) {
+    value_ = std::move(value);
+  }
+
+  c10::optional<T> value_;
+  mutable std::mutex mutex_;
+  mutable std::condition_variable valueCV_;
```

Comment:
Why are these marked as `mutable`?

Reply:
I would like to make `getValue()` function const, but it modifies the mutex and CV. What's the convention here? Should I drop the const modifier on `getValue()` instead?

Reply:
I see, I think it's fine then. @smessmer @ezyang do you have thoughts on this pattern?

Reply:
I follow the standard C++ guidance, which is that 'const'-ness refers to logical constness, and thus mutable is an appropriate use for things like mutexes.